libcxx/include/__memory/unique_ptr.h

```diff
@@ -273,6 +274,9 @@ class _LIBCPP_UNIQUE_PTR_TRIVIAL_ABI _LIBCPP_TEMPLATE_VIS unique_ptr {
   _LIBCPP_HIDE_FROM_ABI _LIBCPP_CONSTEXPR_SINCE_CXX23 void swap(unique_ptr& __u) _NOEXCEPT { __ptr_.swap(__u.__ptr_); }
 };
 
+template <class _Tp>
+struct __libcpp_is_trivially_relocatable<unique_ptr<_Tp>> : true_type {};
+
```

Comment:
Also, I think we might need to care about user-defined specializations, since they are technically allowed. I might have been okay with saying "meh, let's ignore them", but consider something like this:

```c++
template <class T>
struct relative { T val; };

template <class T>
class std::unique_ptr<relative<T>> {
public:
    // same as unique_ptr

private:
    std::ptrdiff_t offset_from_this_;
};

std::unique_ptr<relative<int>> p = std::make_unique<relative<int>>({42});
// p can be stored in a memory mapped file
```

Is this a good idea? I don't know, but that's not too far fetched of a use case, and I'm pretty sure it's valid. However it would break *horribly* if we assume that this is trivially relocatable.

So I think we need to use the `__primary_template` here too.

Reply:
Another way to achieve this is to use a member typedef like BSL does:
```c++
template<class _Tp, class _Dp>
class unique_ptr {
public:
  using __libcpp_trivially_relocatable_tag = conditional_t<
    __libcpp_is_trivially_relocatable<_Dp>::value && __libcpp_is_trivially_relocatable<typename _Dp::pointer>::value,
    unique_ptr, void>;
};
```
and then instead of specializing `__libcpp_is_trivially_relocatable` you could simply make the primary template look like this:
```c++
template <class _Tp, class _Enable = void>
struct __libcpp_is_trivially_relocatable : is_trivially_copyable<_Tp> {};

template <class _Tp>
struct __libcpp_is_trivially_relocatable<_Tp, __enable_if_t<is_same_v<typename _Tp::__libcpp_trivially_relocatable_tag, _Tp>>> : true_type {};
```
Then you wouldn't need to add specializations of `__libcpp_is_trivially_relocatable` in any other headers.